feat: Add input support for EBU Teletext in MPEG-TS

cosmin opened the pull request:
Replaces #1181

* Add support for EBU Teletext input following Level 1.5 of the core specification ETSI EN 300 706 V1.2.1 (2003-04).
* Add support for webvtt in MP4 segments output.

Closes #272

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+    uint8_t data_unit_length;
+    RCHECK(reader.ReadBits(8, &data_unit_length));
+
+    if (data_unit_length != 44) {
```
This would make another good named constant

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+    }
+
+    if (data_unit_id != EBU_TELETEXT_WITH_SUBTITLING) {
+      RCHECK(reader.SkipBytes(44));
```
Use it here, too

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+         4 * bit(address_bits, 4) + 8 * bit(address_bits, 2) +
+         16 * bit(address_bits, 0));
+    const uint8_t* data_block = reader.current_byte_ptr();
+    RCHECK(reader.SkipBytes(40));
```
kConstantFor44 - 4?

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+
+    page_number_ = page_number;
+    magazine_ = magazine;
+    if (page_number == 0xFF) {
```
I was thinking this should be another named constant, but now I realize this condition seems impossible.  Each digit from the hamming code is 4 bits, max of 15.  page_number = 10 * tens + units, or a maximum of 15 * 10 + 15 = 165 == 0xA5

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+
+    return false;
+
+  } else if (packet_nr == 26) {
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+    ParsePacket26(data_block);
+    return false;
+
+  } else if (packet_nr > 26) {
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+
+std::string EsParserTeletext::BuildText(const uint8_t* data_block,
+                                        const uint8_t row) const {
+  const size_t payload_len = 40;
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+    char next_char =
+        static_cast<char>(TELETEXT_BITREVERSE_8[data_block[i]] & 0x7f);
+
+    if (next_char < 32) {
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+        static_cast<char>(TELETEXT_BITREVERSE_8[data_block[i]] & 0x7f);
+
+    if (next_char < 32) {
+      next_char = 0x20;
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+    }
+
+    if (leading_spaces) {
+      if (next_char == 0x20) {
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+
+    switch (next_char) {
+      case '&':
+        next_string.append("&amp;");
```
I'm not very familiar with this part of the code.  Why does the text need to be XML-escaped?

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+  return RemoveTrailingSpaces(next_string);
+}
+
+void EsParserTeletext::ParsePacket26(const uint8_t* data_block) {
```
Better name than packet 26?

cosmin replied:
I can't think of a better name, it does look like "packet 26" is how this is referred to in the various docs I've come across. Maybe someone more familiar with Teletext might have a better idea, but we can always rename this later.

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+  auto& replacement_map = page_state_[index].packet_26_replacements;
+
+  uint8_t row = 0;
+  const size_t payload_len = 40;
```
Named constant, please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+  const size_t payload_len = 40;
+
+  std::vector<uint32_t> x26_triplets;
+  x26_triplets.reserve(13);
```
Named constant or arraysize(), please

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+    const uint8_t address = triplet & 0x3f;
+    const uint8_t row_address_group = (address >= 0x28) && (address <= 0x3f);
+
+    if ((mode == 0x4) && (row_address_group == 0x1)) {
+      row = address - 0x28;
+      if (row == 0x0) {
+        row = 0x18;
+      }
+    }
```
Named constants, please

cosmin replied:
on the one hand I agree these need to be constants, on the other hand that might require someone familiar with the spec to come up with some meaningful names for them.

joeyparrish commented on packager/media/formats/mp2t/es_parser_teletext.cc:
```diff
@@ -0,0 +1,518 @@
+      }
+    }
+
+    if (mode >= 0x11 && mode <= 0x1f && row_address_group == 0x1) {
```
Named constants, please

cosmin commented:
I think I addressed all the feedback except for the named constants in `ParsePacket26` that might require having to carefully go through the spec to come up with some meaningful names for.

I propose we merge this as is (on account of it works and has tests) and then can improve it in the future (such as when adding support for other languages).
